utils/src/chunked_polynomial.rs

```diff
@@ -0,0 +1,164 @@
+
+use rayon::prelude::*;
+
+#[derive(Clone)]
+pub struct ChunkedPolynomials<P> {
+    pub polys: Vec<P>,
+    pub degree: usize,
+}
+
+impl<P> Default for ChunkedPolynomials<P> {
+    fn default() -> ChunkedPolynomials<P> {
+        ChunkedPolynomials {
+            polys: vec![],
+            degree: 0,
+        }
+    }
+}
+
+impl<P> ChunkedPolynomials<P> {
+    pub fn add_chunk(&mut self, p: P) {
+        self.polys.push(p)
+    }
+}
+
+impl<F: Field> ChunkedPolynomials<DensePolynomial<F>> {
+    /// This function evaluates polynomial in chunks.
+    pub fn linearize(&self, elm: F) -> Vec<F> {
+        let mut res: Vec<F> = vec![];
+        for poly in self.polys.iter() {
+            let eval = poly.evaluate(&elm);
+            res.push(eval);
+        }
+        res
+    }
+
+    /// Multiplies the chunks of a polynomial with powers of zeta^n to make it of degree n-1.
+    /// For example, if a polynomial can be written `f = f0 + x^n f1 + x^2n f2`
+    /// (where f0, f1, f2 are of degree n-1), then this function returns the new semi-evaluated
+    /// `f'(x) = f0(x) + zeta^n f1(x) + zeta^2n f2(x)`.
+    pub fn compress_polynomial(&self, zeta_n: F) -> DensePolynomial<F> {
+        let mut scale = F::one();
+        let mut coeffs = vec![F::zero(); self.degree];
+
+        for poly in self.polys.iter() {
+            for (coeff, poly_coeff) in coeffs.iter_mut().zip(&poly.coeffs) {
+                *coeff += scale * poly_coeff;
+            }
+
+            scale *= zeta_n;
+        }
+
+        while coeffs.last().map_or(false, |c| c.is_zero()) {
+            coeffs.pop();
+        }
+
+        DensePolynomial { coeffs }
+    }
+}
+enum OptShiftedPolynomial<P> {
+    Unshifted(P),
+    Shifted(P, usize),
+}
+
+/// A formal sum of the form
+/// `s_0 * p_0 + ... s_n * p_n`
+/// where each `s_i` is a scalar and each `p_i` is an optionally shifted polynomial.
+
+///pub struct ChunkedPolynomial<F, P>(Vec<(F, OptShiftedPolynomial<P>)>);
+
+pub struct ScaledChunkedPolynomials<F, P> {
+    scale: Vec<F>,
+    chunked_polynomials: ChunkedPolynomials<OptShiftedPolynomial<P>>,
+}
+
+impl<F, P> Default for ScaledChunkedPolynomials<F, P> {
+    fn default() -> ScaledChunkedPolynomials<F, P> {
+        ScaledChunkedPolynomials {
+            scale: vec![],
+            chunked_polynomials: ChunkedPolynomials::<OptShiftedPolynomial<P>>::default(),
+        }
+    }
+}
+
+impl<F, P> ScaledChunkedPolynomials<F, P> {
+    pub fn add_unshifted(&mut self, scale: F, p: P) {
+        self.scale.push(scale);
+        self.chunked_polynomials
+            .add_chunk(OptShiftedPolynomial::Unshifted(p))
+    }
+
+    pub fn add_shifted(&mut self, scale: F, shift: usize, p: P) {
+        self.scale.push(scale);
+        self.chunked_polynomials
+            .add_chunk(OptShiftedPolynomial::Shifted(p, shift))
+    }
+}
+
+impl<'a, F: Field> ScaledChunkedPolynomials<F, &'a [F]> {
+    /// check length?
```

Comment:
if this is a TODO, change to normal comment // 
and if it is indeed a TODO, is there any test to check the length? it seems like an important check to make sure dense polynomials will be correctly formed